### ISSUE TITLE
Aputure lights or Sidus Link app

### ISSUE THREAD

**Issue by QNCthomas** [bitfocus/companion-module-requests#623]
**Describe the feature**
Looking to control Aputure lighting via streamdeck with color presets or effects.    Lights such as the Amaran MC series.  The sidus link app allows for presets, and full adjustments.   I don't see an API though to the app, plus it's  phone or tablet based.

**Usecases**
Changing mood lighting or switching scenes or highlights using any aputure lighting setup.

**Comment by chron0** [comment]
https://fccid.io/2A96J-SLT8232 might give us some clues...

i could try to dump BLE traffic and GATTs on my amaran 200x if someone would clue me in what I specifically need to look for.

**Comment by kissu** [comment]
I've seen that tweet reply: https://twitter.com/AputureLighting/status/1610359184985579520
Sent them an email because I find it borderline to promise something and not deliver on it.
The post is also a year old. If they do not wanna release a public API or something, just do not promise it.

**Comment by nicksimoneau** [comment]
Any update on this case? I have some 200 and 300 light I would like to control from API or PC

**Comment by kissu** [comment]
Not gonna happen anytime soon I guess. Faster to reverse engineer it at this point tbh... 😒

**Comment by marnovo** [comment]
Reverse engineering bluetooth lights is indeed a thing, not sure how feasible in this particular case.

- How To Reverse Engineer A Bluetooth Device - YouTube: https://www.youtube.com/watch?v=NIBmiPtCDdM
- Building a new Home Assistant integration in Python - YouTube: https://www.youtube.com/watch?v=e3VwPb72Bbg

Aputure last year said they were considering a PC app for this year, and recently when announcing Sidus Link Pro and Sidus One, have mentioned PC control for Sidus Link lights again, and about other forms of interoperability "not at release".

https://www.reddit.com/r/cinematography/comments/1bzz7wq/aputure_release_sidus_link_pro_and_sidus_one/

My questions and their answers:
1. Will Sidus Link continue and stay BT only, and SLP will be BT + DMX?
**Yes.**
3. Is it going to be available to PC/Mac desktops?
**That is part of the roadmap**
4. Any expectation of interoperability with physical controls of any sort? Remotes, DMX controllers, MIDI controllers, Stream Deck…? Or some sort of software control/API.
**For Sidus link Pro? Not at release.**
5. Will this app be able to keep BT and DMX connected lights from "sleeping" (and disconnecting, taking some time to reconnect once app is reopened)?
**DMX connected lights yes since they are externally paired to Sidus One they'll work instantly, you can quit the up and as soon as the scene loads you have full control of all your lights. Bluetooth mesh lights will face the same limitations and reconnect tab as experienced with Sidus Link.**

**Comment by kissu** [comment]
Hm, got a **no**. Nice.

https://www.reddit.com/r/cinematography/comments/1bzz7wq/comment/l9267sq/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by AliasDjimy** [comment]
Damn Bro ! ..... When is this API coming out ? the SLP app seems promising and the App works on Mac ... all is left is to iiiintegrate it with Bitfocus Companion app & StreamDecks ... Hope They do so before this year is over ...

#SidusLink #SidusOne #bitfocus #companion

**Comment by kissu** [comment]
@AliasDjimy they will probably not do it anytime soon. Even 2025 doesn't sound reasonable considering how silent and low-caring they are about the whole subject.

**Comment by kissu** [comment]
Alright, here we go! Not sure why or how but we got something. 🎊🎊
https://youtu.be/zwG-uT6kXHo

**Comment by chron0** [comment]
that would be really awesome to finally be able to seamlessly control the lights properly.

**Comment by clarkezone** [comment]
Any updates on companion integration for Aputure?  Is anyone working on it?

**Comment by marnovo** [comment]
I have spent little time on this, but findings from the released Apps + SD plugin so far:

Plenty of app versions available here across platforms: https://amarancreators.com/pages/amaran-app-download

I guess an easy starting point is to reverse engineer the Amaran stream deck plugin, although that would still require the desktop app running in the background as the current plugin requires (so far pretty responsive by the way, seems to use websocket communication). But at least we're not limited by their plugin interface implementation, and could use Bitfocus interface instead.

The stream deck plugin seems like a regular compressed zip with a JS app inside, and the main logic seems inside `bin/plugin.js`. Elgato provides an SDK with documentation and templates here (pretty sure folks here know that backwards): https://docs.elgato.com/sdk/plugins/getting-started

Reverse engineering the desktop app itself is more involved, but now more viable than before. Opening the contents of the .app desktop Mac application reveals something Python & QT based but ultimately compiled to a binary—although some interesting info available in JSON configs, and likely more people will figure out over time.

It could make more viable though to reverse engineer the bluetooth (as I exemplified in [my earlier comment](https://github.com/bitfocus/companion-module-requests/issues/623#issuecomment-2166937646)) given the exposed logic of the plugin, and app assets / JSON, but honestly it might not be necessary anymore with what we have now.

I think that's going to be the baseline we can work with for quite a while, _maybe_ our best bet as next step would be asking to open source part of the app (they won't give away the Sidus part) and the plugin itself. My thoughts why:

Based on their recent Sidus devices and Pro apps released, I believe their strategy is to own the software platform and be "the one to rule them all" bringing _in_ control of devices from their + other brands via DMX, etc. and not get control _out_ of the platform. If they ever provide some sort of API, it will be very pro-centric and handsomely paid.  I think the Amaran app + stream deck is as far as they will go here on giving "third-party" control.

Their competitive moat then becomes the software platform, where they play a role as the more reputable brand which can provide a feature-packed, mature, secure (?) and stable platform, very valuable to customers who then become locked in—as the alternative is a myriad of random new brands with amateurish half-baked apps.

E.g.: on the budget end, Neewer gets hardware out every few weeks and nowadays even fairly good; on the prosumer end, Godox is catching up, Zhiyun is bringing awesome hardware at good price. All have apps, but very basic and amateurish. Neewer seems the most committed to software and app/wireless/BT/SD control.

This way they aren't as pressured in margins from hardware manufacturing, something with longer and more expensive development cycles since the budget/prosumer lighting hardware market itself is getting pretty competitive with constant innovation.

**Comment by quinnqiusidus** [comment]
If you have any API and other needs, you can provide them to me. We have some plans in this regard.

**Comment by zsprackett** [comment]
@marnovo @quinnqiusidus Here's an abstraction for the websocket API.  You do need to have the amaran Desktop application open.

https://gist.github.com/zsprackett/29334b9be1e2bd90c1737bd0ba0eaf5c

**Comment by lukashaider** [comment]
@zsprackett works great! Only Issue I was having is controlling multiple lights independently at the same time. Only a single light or a group can be controlled at once.

**Comment by artyfarty** [comment]
@marnovo i tried to reverse Sidus Link blutooth captures some time ago, it uses BLE mesh which is more complicated than simple bluetooth lights, and there's not much info on how to reverse and replay those.

**Comment by quinnqiusidus** [comment]
@artyfarty you can use amaran Desktop application, then you can apply for openapi key, and then you can DIY control

**Comment by nicksimoneau** [comment]
> [@artyfarty](https://github.com/artyfarty) you can use amaran Desktop application, then you can apply for openapi key, and then you can DIY control

What is open api key?!

**Comment by artyfarty** [comment]
@quinnqiusidus yeah, the app seems to work. I'm considering going this route, however i want to integrate it with HASS in the end, and hass runs on a linux machine. I can forward requests from hass to a machine with amaran app running, but this will be a bit janky.

**Comment by quinnqiusidus** [comment]
> [@quinnqiusidus](https://github.com/quinnqiusidus) yeah, the app seems to work. I'm considering going this route, however i want to integrate it with HASS in the end, and hass runs on a linux machine. I can forward requests from hass to a machine with amaran app running, but this will be a bit janky.

@artyfarty 
First, please update amaran Desktop to the latest version (1.0.30):
https://amarancreators.com/pages/amaran-app-download

We have a feedback portal on amaran Desktop, where you can report your problems to us, and our technical support staff will respond to you in a timely manner.

**Comment by marnovo** [comment]
@quinnqiusidus are there release notes for this and/or future versions of the desktop app? Thanks.

**Comment by quinnqiusidus** [comment]
@marnovo 
1. Added Boost mode (compatible with amaran Ace 25x/c)
2. Added color temperature expansion function (compatible with Amaran new product function)
3. Adapted to green product adjustment
4. Color temperature range supports up to 20000K
5. Added sorting function to device list (by name, by online and offline status, by product model)
6. Open API, you can apply for API through Amaran Desktop
7. New version feedback and suggestions support feedback and reply in APP
8. Add application to system notification bar
9. Added auto-start setting

**Comment by thedist** [comment]
Please keep in mind this is a Companion Module Request thread, not a support thread/patch notes for other 3rd party software.

**Comment by nicksimoneau** [comment]
Does the module exist for companion? I'm in to test it!

**Comment by quinnqiusidus** [comment]
> Damn Bro ! ..... When is this API coming out ? the SLP app seems promising and the App works on Mac ... all is left is to iiiintegrate it with Bitfocus Companion app & StreamDecks ... Hope They do so before this year is over ...
> 
> #SidusLink #SidusOne #bitfocus #companion

When is this API coming out ? 
----- 
it's coming now......Welcome to try it.

**Comment by nicksimoneau** [comment]
I have apply to have the api key but still waiting for it

**Comment by nqztv** [comment]
The Sidus Link OpenAPI Usage Guide (https://tools.sidus.link/openapi/docs/usage) describes how to use WebSocket communication to control the lights. You probably need the Amaran Desktop App or something open to receive the communication. I was able to get a Python script up that controlled my lights.  

Can we get the "Missing Documentation" and "Stale" label removed so we can get some progress on getting companion to control Aputure/Amaran lights?

**Comment by ZoneMix** [comment]
I wish I found this issue prior to trying to reverse engineer the Amaran BT traffic...got all the way to figuring out how it is sending commands through the BT channel, but seems like there is encryption set up that is established at the BT link layer...so don't think you would be able to find the encryption key used to decrypt the traffic. Just applied for an API key through the app.

**Comment by marcus-universe** [comment]
> The Sidus Link OpenAPI Usage Guide (https://tools.sidus.link/openapi/docs/usage) describes how to use WebSocket communication to control the lights. You probably need the Amaran Desktop App or something open to receive the communication. I was able to get a Python script up that controlled my lights.
> 
> Can we get the "Missing Documentation" and "Stale" label removed so we can get some progress on getting companion to control Aputure/Amaran lights?

@nqztv Are you using the **Amaran Desktop App** as a relay or communication layer between your Python script and the lights?

I ask because I have multiple **Amaran 200x** lights that can only be controlled via Bluetooth, and I'm trying to find a way to control them with Home Assistant or Companion. Do I need to enable something within the Amaran Desktop App for the OpenAPI to work, or is there anything else I should consider? 

If you don't mind, could you possibly share your Python script? 🙏

**Comment by nqztv** [comment]
@marcus-universe Yes, I have the Amaran Desktop App open and make sure i can turn lights on and off from the app. When you open Amaran Desktop, it should ask you to login or to sign up. Once you have the application open and are logged in, you can click on the icon on the top-right that is a blue 'a' in a white circle. This should open up a new modal. In the new modal, select the "Open API" tab on the left (see https://tools.sidus.link/openapi/docs/apply). Apply for an API key. **_It took me a day or two to get an API key_**. In a couple days, you should be able to go back to the Open API tab and instead of asking you to apply for an API key, it will list your API key which you can use in your Python scripts.

You can see the sample Python app provided by Amaran at: https://tools.sidus.link/openapi/docs/usage#python-example

Install requisite libraries with something like `pip install -U websockets cryptography` 

Replace the `api_secret_key` in the sample app with the API key you see in Amaran Desktop (or reference it from an environmental variable so you don't hard code your API key).

Replace the `request_message` variable with the message you want to send to the lights (see https://tools.sidus.link/openapi/docs/protocol)

First thing you probably want to do is get the node IDs for each of your lights with something like: https://tools.sidus.link/openapi/docs/protocol#get_device_list

Then you can set your lights with a command like: https://tools.sidus.link/openapi/docs/protocol#set_hsi

That should get you started!

**Comment by matthewpurcell** [comment]
Hi Folks,

I know this is a bit late to the party, but I have released a Companion module to control Amaran Desktop.

It is published and available now in Companion 4.0.0+ under the name: **Aputure: Amaran Lights**

You can find full details and features here: https://github.com/bitfocus/companion-module-amaran-desktop

Importantly, this runs entirely locally and does **not** require an Open API key from Aputure.

All constructive feedback on the module is welcome :) Have fun!
